## EDA for 'labeled_anomalies.csv'

In [1]:
import pandas as pd

anoms = pd.read_csv('C:/Users/House/OneDrive/Documents/Sohan_Arun/BTH Course Work/Deep Learning/Assignment 2/data/labeled_anomalies.csv')
display(anoms.head())


chan_id spacecraft                           anomaly_sequences  \
0     P-1       SMAP  [[2149, 2349], [4536, 4844], [3539, 3779]]   
1     S-1       SMAP                              [[5300, 5747]]   
2     E-1       SMAP                [[5000, 5030], [5610, 6086]]   
3     E-2       SMAP                              [[5598, 6995]]   
4     E-3       SMAP                              [[5094, 8306]]   

                                  class  num_values  
0  [contextual, contextual, contextual]        8505  
1                               [point]        7331  
2              [contextual, contextual]        8516  
3                               [point]        8532  
4                               [point]        8307

In [2]:
print("\nDataframe info:")
anoms.info()


Dataframe info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82 entries, 0 to 81
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   chan_id            82 non-null     object
 1   spacecraft         82 non-null     object
 2   anomaly_sequences  82 non-null     object
 3   class              82 non-null     object
 4   num_values         82 non-null     int64 
dtypes: int64(1), object(4)
memory usage: 3.3+ KB


In [3]:
print("Null counts:")
display(anoms.isnull().sum())

Null counts:


chan_id              0
spacecraft           0
anomaly_sequences    0
class                0
num_values           0
dtype: int64

In [5]:
display(anoms.describe(include='all'))

chan_id spacecraft anomaly_sequences    class   num_values
count       82         82                82       82    82.000000
unique      81          2                80        8          NaN
top        P-2       SMAP    [[1250, 1500]]  [point]          NaN
freq         2         55                 2       47          NaN
mean       NaN        NaN               NaN      NaN  6314.195122
std        NaN        NaN               NaN      NaN  2701.236724
min        NaN        NaN               NaN      NaN  1096.000000
25%        NaN        NaN               NaN      NaN  2997.500000
50%        NaN        NaN               NaN      NaN  7895.500000
75%        NaN        NaN               NaN      NaN  8472.000000
max        NaN        NaN               NaN      NaN  8640.000000

In [6]:
for col in anoms.columns:
    print(f"\nColumn: {col}")
    print(anoms[col].value_counts(dropna=False).head(10))


Column: chan_id
chan_id
P-2     2
P-1     1
A-9     1
T-4     1
P-10    1
S-2     1
M-2     1
M-1     1
M-6     1
F-3     1
Name: count, dtype: int64

Column: spacecraft
spacecraft
SMAP    55
MSL     27
Name: count, dtype: int64

Column: anomaly_sequences
anomaly_sequences
[[1250, 1500]]                                2
[[1110, 2250]]                                2
[[2149, 2349], [4536, 4844], [3539, 3779]]    1
[[1172, 1240]]                                1
[[900, 910]]                                  1
[[1850, 2030]]                                1
[[5600, 5640]]                                1
[[4569, 8433]]                                1
[[4569, 8374]]                                1
[[5300, 6420]]                                1
Name: count, dtype: int64

Column: class
class
[point]                                 47
[contextual]                            17
[contextual, contextual]                 6
[point, point]                           4
[contextual, contextual, c

## EDA for Train data 

In [7]:
import numpy as np

data = np.load(r'C:\Users\House\OneDrive\Documents\Sohan_Arun\BTH Course Work\Deep Learning\Assignment 2\data\train\A-1.npy')

df_train = pd.DataFrame(data)

print(df_train.shape) 

(2880, 25)


In [8]:
print(df_train.head())  

      0    1    2    3    4    5    6    7    8    9   ...   15   16   17  \
0  0.999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1  0.999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2  0.999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3  0.999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4  0.999  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    18   19   20   21   22   23   24  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 25 columns]


In [9]:
df_train.describe().T

count      mean           std    min    25%    50%    75%    max
0   2880.0  0.999000  1.110416e-16  0.999  0.999  0.999  0.999  0.999
1   2880.0  0.036111  1.865990e-01  0.000  0.000  0.000  0.000  1.000
2   2880.0  0.009375  9.638638e-02  0.000  0.000  0.000  0.000  1.000
3   2880.0  0.028472  1.663465e-01  0.000  0.000  0.000  0.000  1.000
4   2880.0  0.003472  5.883339e-02  0.000  0.000  0.000  0.000  1.000
5   2880.0  0.170833  3.764286e-01  0.000  0.000  0.000  0.000  1.000
6   2880.0  0.139236  3.462529e-01  0.000  0.000  0.000  0.000  1.000
7   2880.0  0.001389  3.724838e-02  0.000  0.000  0.000  0.000  1.000
8   2880.0  0.000347  1.863390e-02  0.000  0.000  0.000  0.000  1.000
9   2880.0  0.000694  2.634774e-02  0.000  0.000  0.000  0.000  1.000
10  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000
11  2880.0  0.001389  3.724838e-02  0.000  0.000  0.000  0.000  1.000
12  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000
13  2880.0  0.001042  3.226365e-02  0.000  0.000  0.000  0.000  1.000
14  2880.0  0.001042  3.226365e-02  0.000  0.000  0.000  0.000  1.000
15  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000
16  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000
17  2880.0  0.028472  1.663465e-01  0.000  0.000  0.000  0.000  1.000
18  2880.0  0.028472  1.663465e-01  0.000  0.000  0.000  0.000  1.000
19  2880.0  0.014236  1.184834e-01  0.000  0.000  0.000  0.000  1.000
20  2880.0  0.001736  4.163771e-02  0.000  0.000  0.000  0.000  1.000
21  2880.0  0.038542  1.925333e-01  0.000  0.000  0.000  0.000  1.000
22  2880.0  0.037500  1.900165e-01  0.000  0.000  0.000  0.000  1.000
23  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000
24  2880.0  0.000000  0.000000e+00  0.000  0.000  0.000  0.000  0.000

In [10]:
df_train.isnull().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
20    0
21    0
22    0
23    0
24    0
dtype: int64

### Premiliary Notes



- Contains .npy files (NumPy array format)
- Each file is named after a specific telemetry channel
- Files contain T×25 float32 arrays (where T is the number of timestamps)
-25 features (columns col_0 through col_24) over 2,880 time steps.
- Data is already min-max scaled to [0,1] range
- Data is already preprocessed and normalized